srk/src/terminationDTA.ml

```diff
@@ -0,0 +1,586 @@
+  let new_simulation_mat =
+    let simulation_mat =
+      Array.to_list simulation
+      |> List.map (Linear.linterm_of srk)
+      |> QQMatrix.of_rows
+    in
+    let int_eigenspace = QQMatrix.of_rows int_eig_vecs in
+    let int_eigenspace_sim =
+      Linear.QQVectorSpace.simplify
+        (Linear.QQVectorSpace.of_matrix (QQMatrix.mul int_eigenspace simulation_mat))
+      |> QQMatrix.of_rows
+    in
+    match Linear.divide_right int_eigenspace_sim simulation_mat with
+    | Some mS' -> mS'   (* simpl(ES) = mS' * mS*)
+    | None -> assert false
+  in
+  let dom_constraints_lhs = compose_simulation srk (QQMatrix.of_rows non_int_eig_vecs) simulation in
+  let constraints = mk_and srk (List.map (mk_eq srk (mk_zero srk)) (BatArray.to_list dom_constraints_lhs)) in
+  let new_simulation = compose_simulation srk new_simulation_mat simulation in
+  (* new dynamics matrix N, rep matrix M, and simulation matrix S satisfy N S = S M *)
+  let sm = QQMatrix.mul new_simulation_mat tr_matrix in 
+  match Linear.divide_right sm new_simulation_mat with 
+    None -> assert false 
+  | Some mat -> constraints, mat, new_simulation
+
+(* 
+   Characteristic sequence related operations 
+
+   A characteristic sequence S is a periodic sequence of formulas that encodes the 
+   long-time dynamics of an LIA formula G with respect to dynamics matrix A. 
+   Specifically, we have that for k sufficiently large, G(A^k x) = S_k.
+*)
+module XSeq = struct
+
+  let seq_of_true srk =
+    Periodic.make [mk_true srk]
+
+  let seq_of_false srk =
+    Periodic.make [mk_false srk]
+    
+  (* let seq_and srk x y = 
+    Periodic.map2 (fun a b -> mk_and srk [a; b]) x y *)
+
+  let seq_and srk xs =
+    Periodic.mapn (mk_and srk) xs
+  
+  (* let seq_or srk x y =
+    Periodic.map2 (fun a b -> mk_or srk [a; b]) x y *)
+
+  let seq_or srk xs =
+    Periodic.mapn (mk_or srk) xs
+
+  let seq_add_term srk x y =
+    Periodic.map2 (fun a b -> mk_add srk [a; b]) x y
+
+  let seq_add_int x y =
+    Periodic.map2 (fun a b -> a + b) x y
+
+  let seq_mul_symbol srk x symbol =
+    Periodic.map (fun t -> mk_mul srk [mk_int srk t; mk_const srk symbol]) x
+
+  let seq_not srk x = 
+    Periodic.map (fun f -> mk_not srk f) x
+
+  (* seq_of_exp m t returns the characteristic sequence of t^k mod m, 
+  k = 0, 1, 2, ... *)
+  let seq_of_exp modulus lambda = 
+    UltimatelyPeriodic.unfold (fun power -> (power * lambda) mod modulus) 1 
+    |> periodic_approx
+
+
+  (* seq_of_polynomial srk m p returns characteristic sequence of p(k) mod m, 
+  k = 0, 1, 2, ... *)
+  let seq_of_polynomial modulus poly = 
+    let lcm_of_denoms =
+      BatEnum.fold (fun current_lcm (coeff, _) -> ZZ.lcm current_lcm (QQ.denominator coeff)) 
+      ZZ.one
+      (Polynomial.QQX.enum poly)
+    in
+    let poly = Polynomial.QQX.scalar_mul (QQ.of_zz lcm_of_denoms) poly in
+    let m = modulus * (BatOption.get (ZZ.to_int lcm_of_denoms)) in
+    (0 -- (modulus - 1))
```

Comment:
Should be m?  (Style: I would do "let modulus = modulus * ... in ... " to make sure I don't re-use the (incorrect) modulus).